lib/sms/sms_status_report.c

```diff
@@ -0,0 +1,386 @@
+
+	/* 2 for length and type fields */
+	return 2 + length;
+}
+
+/**
+ * @brief Decode TP-Service-Centre-Time-Stamp as specified in 3GPP TS 23.040 Section 9.2.3.11.
+ *
+ * @param[in,out] parser Parser instance.
+ * @param[in] buf Buffer containing PDU and pointing to this field.
+ *
+ * @return Number of parsed bytes.
+ */
+static int parse_pdu_scts_field(struct parser *parser, uint8_t *buf)
+{
+	struct pdu_status_report_data * const pdata = parser->data;
+	int tmp_tz;
+
+	pdata->scts.year   = semioctet_to_dec(*(buf++));
+	pdata->scts.month  = semioctet_to_dec(*(buf++));
+	pdata->scts.day    = semioctet_to_dec(*(buf++));
+	pdata->scts.hour   = semioctet_to_dec(*(buf++));
+	pdata->scts.minute = semioctet_to_dec(*(buf++));
+	pdata->scts.second = semioctet_to_dec(*(buf++));
+
+	/* 3GPP TS 23.040 Section 9.2.3.11:
+	 *  "The Time Zone indicates the difference, expressed in quarters of an hour,
+	 *   between the local time and GMT. In the first of the two semi octets, the first bit
+	 *   (bit 3 of the seventh octet of the TP Service Centre Time Stamp field) represents
+	 *   the algebraic sign of this difference (0: positive, 1: negative)."
+	 */
+	tmp_tz = ((*buf & 0xf7) * 15) / 60;
+
+	if (*buf & 0x08) {
+		tmp_tz = -(tmp_tz);
+	}
+
+	pdata->scts.timezone = tmp_tz;
+
+	return SCTS_FIELD_SIZE;
+}
+
+/**
+ * @brief Decode TP-Discharge-Time-Stamp as specified in 3GPP TS 23.040 Section 9.2.3.13.
+ *
+ * @param[in,out] parser Parser instance.
+ * @param[in] buf Buffer containing PDU and pointing to this field.
+ *
+ * @return Number of parsed bytes.
+ */
+static int parse_pdu_dt_field(struct parser *parser, uint8_t *buf)
+{
+	struct pdu_status_report_data * const pdata = parser->data;
+	int tmp_tz;
+
+	pdata->dt.year   = semioctet_to_dec(*(buf++));
+	pdata->dt.month  = semioctet_to_dec(*(buf++));
+	pdata->dt.day    = semioctet_to_dec(*(buf++));
+	pdata->dt.hour   = semioctet_to_dec(*(buf++));
+	pdata->dt.minute = semioctet_to_dec(*(buf++));
+	pdata->dt.second = semioctet_to_dec(*(buf++));
+
+	/* 3GPP TS 23.040 Section 9.2.3.11:
+	 *  "The Time Zone indicates the difference, expressed in quarters of an hour,
+	 *   between the local time and GMT. In the first of the two semi octets, the first bit
+	 *   (bit 3 of the seventh octet of the TP Service Centre Time Stamp field) represents
+	 *   the algebraic sign of this difference (0: positive, 1: negative)."
+	 */
+	tmp_tz = ((*buf & 0xf7) * 15) / 60;
+
+	if (*buf & 0x08) {
+		tmp_tz = -(tmp_tz);
+	}
+
+	pdata->dt.timezone = tmp_tz;
+
+	return DT_FIELD_SIZE;
+}
+
+/**
+ * @brief Decode TP-Status as specified in 3GPP TS 23.040 Section 9.2.3.15.
+ *
+ * @param[in,out] parser Parser instance.
+ * @param[in] buf Buffer containing PDU and pointing to this field.
+ *
+ * @return Number of parsed bytes.
+ */
+static int parse_pdu_st_field(struct parser *parser, uint8_t *buf)
+{
+	struct pdu_status_report_data * const pdata = parser->data;
+
+	pdata->st = *((struct pdu_status_report_status *)buf);
```

Comment:
If bit#7 is 1, setting the status of first 7 bits (`status` variable) is not valid. I don't know how to solve that in the best way. Failing parsing sounds too dramatic also.

Reply:
Closing this comment as SMS library changes have been removed from this PR.